handling-forms/controlled-forms.md

```diff
@@ -0,0 +1,542 @@
+        name: '',
+        email: ''
+    });
+```
+<!-- prettier-ignore-end -->
+
+Now she needs to make the input fields read from this state. Since `formFields` is an object, we'll use dot-notation to access the `name` and `email` values.
+
+<!-- prettier-ignore-start -->
+```js
+    return (
+        <form>
+            <div>
+                <label htmlFor="fullName">Name:</label>
+                <input
+                    name="fullName"
+                    value={formFields.name}/>
+            </div>
+            <div>
+                <label htmlFor="email">Email:</label>
+                <input
+                    name="email"
+                    value={formFields.email}/>
+            </div>
+            <input
+                type="submit"
+                value="Add Student"
+            />
+        </form>
+    );
+```
+<!-- prettier-ignore-end -->
+
+### Event Handling To Update `formFields`
+
+Sofia's next step is to ensure that changes in an input field affect the component state.
+
+She can create two event handlers, `onNameChange` and `onEmailChange`. These event handlers need to do two things:
+
+- Read the current value inside the input field
+- Update state to that current value
+
+<!-- prettier-ignore-start -->
+```js
+    const onNameChange = (event) => {
+        setFormFields({
+            ...formFields,
+            name: event.target.value
+        })
+    };
+
+    const onEmailChange = (event) => {
+        setFormFields({
+            ...formFields,
+            email: event.target.value
+        })
+    };
+```
+<!-- prettier-ignore-end -->
+
+In this example, we're using [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals) for a quick way to clone the original `formFields` object. Each event handler should add a specific key-value pair: `onNameChange` adds the key-value pair `name: event.target.value`, while `onEmailChange` adds `email`.
+
+Now, Sofia needs to make sure that the input fields use these event handlers when they're changed:
+
+<!-- prettier-ignore-start -->
+```js
+    return (
+        <form>
+            <div>
+                <label htmlFor="fullName">Name:</label>
+                <input
+                    name="fullName"
+                    value={formFields.name}
+                    onChange={onNameChange}/>
+            </div>
+            <div>
+                <label htmlFor="email">Email:</label>
+                <input name="email"
+                    value={formFields.email}
+                    onChange={onEmailChange}/>
+            </div>
+            <input
+                type="submit"
+                value="Add Student"
+            />
+        </form>
+    );
+```
+<!-- prettier-ignore-end -->
+
+We did it! 🎉🎉
+
+Sofia can verify that her form works as expected:
+
+![](../assets/handling-forms_controlled-forms_newstudentform-dev-tools.png)
+
```

Comment:
I'd really like to bring the in-progress all-in-one view of the code into the lesson here. I know it's also at the start of the next lesson, but it feels really strange that we don't have a spot int _this_ lesson that makes the full code easily available.
No suggestion because code
markdown in following comment

Reply:
### !callout-info

## The `NewStudentForm` Source

Sofia's in-progress version of the `NewStudentForm` is available here.

<br/>

<details>

<summary>A <code>NewStudentForm</code> component, responsible for containing a form to add a new student.</summary>

<!-- prettier-ignore-start -->
```js
import { useState } from 'react';

const NewStudentForm = () => {

    const [formFields, setFormFields] = useState({
        name: '',
        email: ''
    });

    const onNameChange = (event) => {
        setFormFields({
            ...formFields,
            name: event.target.value
        })
    };

    const onEmailChange = (event) => {
        setFormFields({
            ...formFields,
            email: event.target.value
        })
    };

    return (
        <form>
            <div>
                <label htmlFor="fullName">Name:</label>
                <input
                    name="fullName"
                    value={formFields.name}
                    onChange={onNameChange} />
            </div>
            <div>
                <label htmlFor="email">Email:</label>
                <input name="email"
                    value={formFields.email}
                    onChange={onEmailChange} />
            </div>
            <input
                type="submit"
                value="Add Student" />
        </form>
    );
};

export default NewStudentForm;
```
<!-- prettier-ignore-end -->

</details>

### !end-callout